File: leetcode/0001-two-sum.js
// hi
/**  hash map
 * time complexity: O(n)
 * space complexity: O(n)
 */

const twoSum = (n, t) => {
  let hash = {};
  for (let i in n) {
    if (hash[n[i]]) {
      return [i, hash[n[i]]];
    } else {
      hash[t - n[i]] = i;
    }
  }
};

/** nested for
 * time complexity: O(n2)
 * space complexity: O(1)
 */

var twoSum = function (n, t) {
  for (i = 0; i < n.length; i++) {
    for (j = i + 1; j < n.length; j++) {
      if (n[i] + n[j] == t) {
        return [i, j];
      }
    }
  }
};

console.log(twoSum([2, 7, 11, 15], 9), twoSum([3, 2, 4], 6), twoSum([3, 3], 6));

File: leetcode/0001-two-sum.py
# TC: O(1) - hashmap lookup
# utilize hashmap

def twoSum(self, nums, target):
    dict = {}
    # sum megapost sol - https://tinyurl.com/yttmf4e4
    for idx, ele in enumerate(nums):
        remain = target - ele
        if remain in dict:
            return [dict[remain], idx]
        else:
            dict[ele] = idx

# TC: O N^2
# nested for loop
# def twoSum(nums, target):
#     for i in range(len(nums)-1):
#         for j in range(i+1, len(nums)):
#             if(nums[i]+nums[j] == target):
#                 return [i, j]


print(twoSum([2, 7, 11, 15], 9))
